### ISSUE TITLE
Using `insertUI()` with a leaflet widget containing a polygon breaks shiny

### ISSUE THREAD

**Issue by sgrubsmyon** [rstudio/leaflet#418]
Hi,

I am building a web app with Shiny whose central component is the dynamic insertion of content into an infinite chain. A user can select what they want to see, then click a button to display that content. New content is put at the top of a scroll area. Old content slides down when new content is added.

This has the advantage that it is easy and transparent to open any kind of visualization next to each other (actually on top of each other), allowing for easy comparison of content. It is also very flexible since the number of visualizations is not fixed.

The problem I am facing is that some htmlwidgets (e.g. leaflet) cause Shiny to crash when they are inserted with `insertUI()`. No other content (e.g. plot) can be inserted afterwards, if the htmlwidget remains in the DOM. If the htmlwidget is first removed, then content can be inserted again. This happens only under some circumstances, e.g. when a polygon is displayed.

I have created a minimal working example (MWE) to demonstrate.

Consider the following app, which has two buttons to insert a plot and a leaflet map (with polygon).

```r
library(shiny)
library(leaflet)

ui <- fluidPage(
  sidebarLayout(
    sidebarPanel(
      p(paste("Click 'Show Plot', then 'Show Map', then again 'Show Plot'.",
              "Observe that the second plot is empty.",
              "Also, in your browser's error console, you will likely see the",
              "message 'Error: All fields must be arrays'.")
      ),
      p(paste("If you continue clicking 'Show Plot', you will get",
              "'uncaught exception: Duplicate binding for ID map_1' in the",
              "console, even though there is only one object with ID 'map_1'.")
      ),
      p(
        actionButton("plot_btn", "Show Plot"),
        actionButton("map_btn", "Show Map")
      )
    ),
    mainPanel(
      div(id = "container")
    )
  )
)

server <- function(input, output) {
  observeEvent(input$plot_btn, {
    id <- paste0("plot_", input$plot_btn)
    output[[id]] <- renderPlot({
      x <- faithful[, 2]
      bins <- seq(min(x), max(x), length.out = 21)
      hist(x, breaks = bins, col = 'darkgray', border = 'white')
    })
    insertUI("#container", "afterBegin", plotOutput(id))
  })

  lng <- c(-94.17, -94.17, -94.14, -94.14)
  lat <- c(36.1, 36, 36, 36.1)

  observeEvent(input$map_btn, {
    id <- paste0("map_", input$map_btn)
    output[[id]] <- renderLeaflet({
      l <- leaflet() %>% addTiles()
      l <- l %>% addPolygons(lng, lat)
      # l <- l %>% addMarkers(lng, lat)
      l
    })
    insertUI("#container", "afterBegin", leafletOutput(id))
  })
}

shinyApp(ui = ui, server = server)
```

When I use the same code, but instead of appending content with `insertUI()`, I always replace the content with `output$container <- renderUI({})`, then it works fine. See following code:

```r
library(shiny)
library(leaflet)

ui <- fluidPage(
  sidebarLayout(
    sidebarPanel(
      p(paste("In this example, it works just fine. The content is not appended",
              "with insertUI(), but replaced with output$container <- renderUI().")
      ),
      p(
        actionButton("plot_btn", "Show Plot"),
        actionButton("map_btn", "Show Map")
      )
    ),
    mainPanel(
      uiOutput("container")
    )
  )
)

server <- function(input, output) {
  observeEvent(input$plot_btn, {
    id <- paste0("plot_", input$plot_btn)
    output[[id]] <- renderPlot({
      x <- faithful[, 2] 
      bins <- seq(min(x), max(x), length.out = 21)
      hist(x, breaks = bins, col = 'darkgray', border = 'white')
    })
    output$container <- renderUI({
      plotOutput(id)
    })
  })
  
  lng <- c(-94.17, -94.17, -94.14, -94.14)
  lat <- c(36.1, 36, 36, 36.1)

  observeEvent(input$map_btn, {
    id <- paste0("map_", input$map_btn)
    output[[id]] <- renderLeaflet({
      l <- leaflet() %>% addTiles()
      l <- l %>% addPolygons(lng, lat)
      # l <- l %>% addMarkers(lng, lat)
      l
    })
    output$container <- renderUI({
      leafletOutput(id)
    })
  })
}

shinyApp(ui = ui, server = server)
```

I have searched a bit to try and understand this behavior. According to @jcheng5's comment [here](https://github.com/MangoTheCat/radarchart/issues/31#issuecomment-234400473), using `insertUI()` to insert charts should work. Surprisingly, it also works using `addMarkers()` instead of `addPolygons()` (you can uncomment in the code to check).

It is beyond me to understand where the problem actually occurs. Note also the errors in the browser console log, e.g. complaining about `Duplicate binding for ID map_1`. In the browser's inspector, you can see that the DOM contains only one HTML object with that ID. I think that there is some JS code running in the back that leads to this weird behavior. Maybe it is related to [this problem](https://stackoverflow.com/questions/20600088/outputbindings-register-throws-error-in-shiny). It seems that somehow leaflet with a polygon is blocking a binding, which would be needed to run reactive shiny code. This is why any other content inserted after the leaflet map just remains an empty div and is not filled with a plot by Shiny's JS bindings.

If someone knows any workaround, I would be very glad to know. Maybe I am misusing htmlwidgets in some way? Is there another, more elegant way?

The only workaround I can think of right now is to remove any widget causing problems with `removeUI()` before a new widget is inserted, which kind of goes against the whole point of my app design, because comparing any content is not possible.

Or, if this task is not possible to achieve with Shiny, I would perhaps choose to use another web app framework (or pure JS only to solve this particular problem). So far, I am very happy with Shiny and htmlwidgets as they allow to write very clean, modular, concise code. Thank you so much for the hard work on R, Shiny and all the other useful R packages!

**Comment by sgrubsmyon** [comment]
I was using version ‘1.0.1’ from CRAN, after watching [the rconf video on leaflet](https://www.rstudio.com/resources/videos/mapping-in-r-with-leaflet/), I upgraded with `devtools::install_github("rstudio/leaflet")`, now I have version ‘1.1.0.9000’ and the issue is gone!

Does anyone know from the top of the head which change in the code solved this? Because I was experiencing a similar problem using the highcharter htmlwidget with choropleth maps. Maybe it's a similar issue that could be fixed by the same measures.

**Comment by bhaskarvk** [comment]
BTW CRAN is now on ver. 1.1.0 too, but 1.1.0.9000 has even more stuff after 1.1.0 was released to CRAN.
I think the problem could've been in the htmlwidget package itself, but I can't be sure. May be when you upgraded leaflet, it also pulled in the latest htmlwidgets package. There was a bug fix in htmlwidget (0.8.0) related to dependencies+shiny. I am just guessing though, so take it w/ a grain of salt.

**Comment by sgrubsmyon** [comment]
Thanks @bhaskarvk for letting me know. That might be. I notices that quite a number of packages was updated as dependencies, e.g. viridisLite, but I did not see if htmlwidgets was among them. I'm not on htmlwidgets version 0.8. First checks indicate that in highcharter the problem is still there, but this is not an issue of leaflet of course.
